analyzers/src/SonarAnalyzer.CSharp/Rules/EqualityOnModulus.cs

```diff
@@ -50,27 +61,59 @@ private static void VisitEquality(SyntaxNodeAnalysisContext c)
-                                                               constantValue < 0 ? "negative" : "positive"));
+                c.ReportIssue(Diagnostic.Create(Rule, equalsExpression.GetLocation(), constantValue < 0 ? "negative" : "positive"));
             }
         }
 
-        private static bool CheckExpression(SyntaxNode node, ExpressionSyntax modulus, SemanticModel semanticModel, out int constantValue) =>
+        private static bool CheckExpression(SyntaxNode node, ExpressionSyntax expression, SemanticModel semanticModel, out int constantValue) =>
             ExpressionNumericConverter.TryGetConstantIntValue(node, out constantValue)
             && constantValue != 0
-            && IsModulus(modulus)
-            && !IsUnsigned(modulus, semanticModel);
-
-        private static bool IsModulus(ExpressionSyntax expression) =>
-            expression.RemoveParentheses() is BinaryExpressionSyntax binary
-            && binary.IsKind(SyntaxKind.ModuloExpression);
+            && expression.RemoveParentheses() is BinaryExpressionSyntax binary
+            && binary.IsKind(SyntaxKind.ModuloExpression)
+            && !ExpressionIsAlwaysPositive(binary, semanticModel);
 
-        private static bool IsUnsigned(ExpressionSyntax expression, SemanticModel semantic)
+        private static bool ExpressionIsAlwaysPositive(BinaryExpressionSyntax binaryExpression, SemanticModel semantic)
         {
-            var type = semantic.GetTypeInfo(expression).Type;
-            return type.IsAny(KnownType.UnsignedIntegers)
-                   || type.Is(KnownType.System_UIntPtr);
+            var type = semantic.GetTypeInfo(binaryExpression).Type;
+            if (type.IsAny(KnownType.UnsignedIntegers) || type.Is(KnownType.System_UIntPtr))
+            {
+                return true;
+            }
+
+            var leftExpression = binaryExpression.Left;
+            var leftExpressionStringForm = leftExpression.ToString();
+            return CollectionSizePropertyOrMethodNames.Any(x => leftExpressionStringForm.Contains(x))
+                   && semantic.GetSymbolInfo(leftExpression).Symbol is { } symbol
+                   && IsCollectionSize(symbol);
         }
+
+        private static bool IsCollectionSize(ISymbol symbol) =>
+            IsEnumerableCountMethod(symbol)
+            || IsLengthProperty(symbol)
+            || IsCollectionCountProperty(symbol)
+            || IsListCapacityProperty(symbol);
+
+        private static bool IsEnumerableCountMethod(ISymbol symbol) =>
+            (CountName.Equals(symbol.Name) || LongCountName.Equals(symbol.Name))
+            && symbol is IMethodSymbol methodSymbol
+            && methodSymbol.IsExtensionMethod
+            && methodSymbol.ReceiverType != null
+            && methodSymbol.IsExtensionOn(KnownType.System_Collections_Generic_IEnumerable_T);
+
+        private static bool IsLengthProperty(ISymbol symbol) =>
```

Comment:
I would  group these three checks IsLengthProperty(symbol) IsCollectionCountProperty(symbol) and IsListCapacityProperty(symbol) into a different method and do the `is IPropertySymbol` check once and use the `PropertySymbol` as an argument for them.

Reply:
NITPICK: I think you could have left the three methods to divide the different types with different properties like: 

```
        private static bool IsLengthProperty(IPropertySymbol propertySymbol) =>
            (LengthName.Equals(propertySymbol.Name) || LongLengthName.Equals(propertySymbol.Name))
            && propertySymbol.ContainingType.IsAny(KnownType.System_Array, KnownType.System_String);

        private static bool IsCollectionCountProperty(IPropertySymbol propertySymbol) =>
            CountName.Equals(propertySymbol.Name)
            && (propertySymbol.ContainingType.Implements(KnownType.System_Collections_Generic_ICollection_T)
                || propertySymbol.ContainingType.Implements(KnownType.System_Collections_Generic_IReadOnlyCollection_T));

        private static bool IsListCapacityProperty(IPropertySymbol propertySymbol) =>
            ListCapacityName.Equals(propertySymbol.Name)
            && propertySymbol.ContainingType.Implements(KnownType.System_Collections_Generic_IList_T);
```